File: apps/android/app/src/main/kotlin/com/edge/marker/app/TrackingEngine.kt
package com.edge.marker.app

import com.edge.marker.domain.DayKey
import com.edge.marker.domain.MarkerWeekday
import com.edge.marker.domain.Tracker
import com.edge.marker.domain.TrackerSchedule
import com.edge.marker.domain.TrackingEntry
import com.edge.marker.domain.UserPreference
import java.util.TimeZone
import java.util.UUID

data class TodayTrackerItem(
    val tracker: Tracker,
    val isCompleted: Boolean,
    val weeklyProgressText: String?,
) {
    val id: UUID
        get() = tracker.id
}

data class HistoryDayItem(
    val trackerName: String,
    val colorToken: String,
    val recordedAtMillis: Long,
) {
    val id: String
        get() = "$trackerName-$recordedAtMillis"
}

data class HistoryDaySection(
    val dayKey: DayKey,
    val items: List<HistoryDayItem>,
) {
    val id: String
        get() = dayKey.rawValue
}

data class TrackerCompletionSummary(
    val trackerId: UUID,
    val trackerName: String,
    val colorToken: String,
    val totalEntryCount: Int,
) {
    val id: UUID
        get() = trackerId
}

data class StatisticsSummary(
    val activeTrackerCount: Int,
    val totalEntryCount: Int,
    val currentWindowCompletedCount: Int,
    val currentWindowDueCount: Int,
    val currentWindowCompletionRate: Double,
    val currentStreakDays: Int,
    val trackerBreakdown: List<TrackerCompletionSummary>,
)

object TrackingEngine {
    fun buildTodayItems(
        trackers: List<Tracker>,
        entries: List<TrackingEntry>,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday,
    ): List<TodayTrackerItem> {
        val activeTrackers: List<Tracker> = trackers.filter { tracker -> !tracker.isArchived }
        val completedKeys: Set<String> = entries.mapTo(mutableSetOf()) { entry ->
            key(trackerId = entry.trackerId, dayKey = entry.dayKey)
        }

        return activeTrackers.mapNotNull { tracker ->
            val completedCount: Int = completionCountBeforeOrOnDay(
                entries = entries,
                trackerId = tracker.id,
                dayKey = dayKey,
                weekStartsOn = weekStartsOn,
            )
            val isDue: Boolean = tracker.schedule.isDue(on = dayKey, completedCountInWeek = completedCount)

            if (!isDue) {
                null
            } else {
                TodayTrackerItem(
                    tracker = tracker,
                    isCompleted = completedKeys.contains(key(trackerId = tracker.id, dayKey = dayKey)),
                    weeklyProgressText = when (val schedule = tracker.schedule) {
                        is TrackerSchedule.WeeklyQuota -> "${minOf(completedCount, schedule.targetCount)}/${schedule.targetCount}"
                        else -> null
                    },
                )
            }
        }.sortedBy { item -> item.tracker.createdAt }
    }

    fun buildHistorySections(
        trackers: List<Tracker>,
        entries: List<TrackingEntry>,
    ): List<HistoryDaySection> {
        val trackerLookup: Map<UUID, Tracker> = trackers.associateBy { tracker -> tracker.id }

        return entries
            .groupBy { entry -> entry.dayKey }
            .toSortedMap(reverseOrder())
            .map { (dayKey, groupedEntries) ->
                HistoryDaySection(
                    dayKey = dayKey,
                    items = groupedEntries
                        .sortedByDescending { entry -> entry.recordedAt.time }
                        .mapNotNull { entry ->
                            trackerLookup[entry.trackerId]?.let { tracker ->
                                HistoryDayItem(
                                    trackerName = tracker.name,
                                    colorToken = tracker.colorToken,
                                    recordedAtMillis = entry.recordedAt.time,
                                )
                            }
                        },
                )
            }
            .filter { section -> section.items.isNotEmpty() }
    }

    fun buildStatisticsSummary(
        trackers: List<Tracker>,
        entries: List<TrackingEntry>,
        today: DayKey,
        preferences: UserPreference,
    ): StatisticsSummary {
        val activeTrackers: List<Tracker> = trackers.filter { tracker -> !tracker.isArchived }
        val completedKeys: Set<String> = entries.mapTo(mutableSetOf()) { entry ->
            key(trackerId = entry.trackerId, dayKey = entry.dayKey)
        }
        val windowStart: DayKey = today.addingDays(days = -(preferences.preferredStatisticsWindow.dayCount - 1))

        var dueCount = 0
        var completedCount = 0
        var currentDay: DayKey = windowStart

        while (currentDay <= today) {
            activeTrackers
                .filter { tracker -> trackerCreatedDayKey(tracker = tracker) <= currentDay }
                .forEach { tracker ->
                    val weeklyCountBeforeDay: Int = completionCountBeforeDay(
                        entries = entries,
                        trackerId = tracker.id,
                        dayKey = currentDay,
                        weekStartsOn = preferences.weekStartsOn,
                    )

                    if (tracker.schedule.isDue(on = currentDay, completedCountInWeek = weeklyCountBeforeDay)) {
                        dueCount += 1
                        if (completedKeys.contains(key(trackerId = tracker.id, dayKey = currentDay))) {
                            completedCount += 1
                        }
                    }
                }

            if (currentDay == today) {
                break
            }
            currentDay = currentDay.addingDays(days = 1)
        }

        val trackerBreakdown: List<TrackerCompletionSummary> = activeTrackers
            .map { tracker ->
                TrackerCompletionSummary(
                    trackerId = tracker.id,
                    trackerName = tracker.name,
                    colorToken = tracker.colorToken,
                    totalEntryCount = entries.count { entry -> entry.trackerId == tracker.id },
                )
            }
            .sortedWith(
                compareByDescending<TrackerCompletionSummary> { summary -> summary.totalEntryCount }
                    .thenBy { summary -> summary.trackerName },
            )

        return StatisticsSummary(
            activeTrackerCount = activeTrackers.count(),
            totalEntryCount = entries.count(),
            currentWindowCompletedCount = completedCount,
            currentWindowDueCount = dueCount,
            currentWindowCompletionRate = if (dueCount == 0) 0.0 else completedCount.toDouble() / dueCount.toDouble(),
            currentStreakDays = buildCurrentStreakDays(entries = entries, today = today),
            trackerBreakdown = trackerBreakdown,
        )
    }

    private fun buildCurrentStreakDays(entries: List<TrackingEntry>, today: DayKey): Int {
        val completedDays: Set<DayKey> = entries.mapTo(mutableSetOf()) { entry -> entry.dayKey }
        var streak = 0
        var cursor: DayKey = today

        while (completedDays.contains(cursor)) {
            streak += 1
            cursor = cursor.addingDays(days = -1)
        }

        return streak
    }

    private fun completionCountBeforeOrOnDay(
        entries: List<TrackingEntry>,
        trackerId: UUID,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday,
    ): Int = completionCount(
        entries = entries,
        trackerId = trackerId,
        dayKey = dayKey,
        weekStartsOn = weekStartsOn,
        includeCurrentDay = true,
    )

    private fun completionCountBeforeDay(
        entries: List<TrackingEntry>,
        trackerId: UUID,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday,
    ): Int = completionCount(
        entries = entries,
        trackerId = trackerId,
        dayKey = dayKey,
        weekStartsOn = weekStartsOn,
        includeCurrentDay = false,
    )

    private fun completionCount(
        entries: List<TrackingEntry>,
        trackerId: UUID,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday,
        includeCurrentDay: Boolean,
    ): Int {
        val bounds: WeekBounds = weekBounds(dayKey = dayKey, weekStartsOn = weekStartsOn)

        return entries.count { entry ->
            entry.trackerId == trackerId &&
                entry.dayKey >= bounds.start &&
                entry.dayKey <= bounds.end &&
                if (includeCurrentDay) entry.dayKey <= dayKey else entry.dayKey < dayKey
        }
    }

    private fun weekBounds(dayKey: DayKey, weekStartsOn: MarkerWeekday): WeekBounds {
        val offset: Int = (dayKey.weekday.rawValue - weekStartsOn.rawValue + 7) % 7
        val start: DayKey = dayKey.addingDays(days = -offset)
        val end: DayKey = start.addingDays(days = 6)
        return WeekBounds(start = start, end = end)
    }

    private fun key(trackerId: UUID, dayKey: DayKey): String = "${trackerId}|${dayKey.rawValue}"

    private fun trackerCreatedDayKey(tracker: Tracker): DayKey = DayKey(
        date = tracker.createdAt,
        timeZone = TimeZone.getDefault(),
    )
}

private data class WeekBounds(
    val start: DayKey,
    val end: DayKey,
)

File: apps/ios/MarkerApp/Sources/App/TrackingEngine.swift
import Foundation
import MarkerDomain

struct TodayTrackerItem: Identifiable {
    let tracker: Tracker
    let entry: TrackingEntry?
    let weeklyProgressText: String?

    var id: UUID { tracker.id }
    var hasRecord: Bool { entry != nil }
    var isCompleted: Bool { entry?.countsAsCompletion ?? false }
    var entrySummaryText: String? { entry?.summary }
    var isSkippedMedication: Bool {
        entry?.payload.kind == .medication && entry?.payload.medicationStatus == .skipped
    }
}

struct TodayOverview {
    let activeTrackerCount: Int
    let pendingItems: [TodayTrackerItem]
    let recordedItems: [TodayTrackerItem]

    var allItems: [TodayTrackerItem] {
        pendingItems + recordedItems
    }

    var summaryText: String {
        if activeTrackerCount == 0 {
            return "先记录一件照顾自己的事"
        }

        if pendingItems.isEmpty && recordedItems.isEmpty {
            return "今天没有需要记录的项目"
        }

        if pendingItems.isEmpty {
            return "今天的记录已保存"
        }

        return "今天还有 \(pendingItems.count) 项待确认"
    }
}

struct HistoryDayItem: Identifiable {
    let trackerName: String
    let colorToken: String
    let recordedAt: Date
    let payloadSummary: String

    var id: String { "\(trackerName)-\(recordedAt.timeIntervalSince1970)" }
}

struct HistoryDaySection: Identifiable {
    let dayKey: DayKey
    let items: [HistoryDayItem]

    var id: String { dayKey.rawValue }
}

struct TrackerCompletionSummary: Identifiable {
    let trackerID: UUID
    let trackerName: String
    let colorToken: String
    let totalEntryCount: Int

    var id: UUID { trackerID }
}

struct StatisticsSummary {
    let activeTrackerCount: Int
    let totalEntryCount: Int
    let currentWindowCompletedCount: Int
    let currentWindowDueCount: Int
    let currentWindowCompletionRate: Double
    let currentStreakDays: Int
    let trackerBreakdown: [TrackerCompletionSummary]
}

enum TrackingEngine {
    static func buildTodayOverview(
        trackers: [Tracker],
        entries: [TrackingEntry],
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday
    ) -> TodayOverview {
        let activeTrackers = trackers.filter { !$0.isArchived }
        let todayEntries = Dictionary(uniqueKeysWithValues: entries
            .filter { $0.dayKey == dayKey }
            .map { ($0.trackerId, $0) })

        var pendingItems: [TodayTrackerItem] = []
        var recordedItems: [TodayTrackerItem] = []

        for tracker in activeTrackers.sorted(by: { $0.createdAt < $1.createdAt }) {
            let todayEntry = todayEntries[tracker.id]
            let weeklyCountBeforeToday = completionCountBeforeDay(
                entries: entries,
                trackerID: tracker.id,
                dayKey: dayKey,
                weekStartsOn: weekStartsOn
            )
            let isDue = tracker.schedule.isDue(on: dayKey, completedCountInWeek: weeklyCountBeforeToday)

            guard isDue || todayEntry != nil else { continue }

            let weeklyProgressText: String?
            if case let .weeklyQuota(targetCount) = tracker.schedule {
                let completedCount = completionCountBeforeOrOnDay(
                    entries: entries,
                    trackerID: tracker.id,
                    dayKey: dayKey,
                    weekStartsOn: weekStartsOn
                )
                weeklyProgressText = "\(min(completedCount, targetCount))/\(targetCount)"
            } else {
                weeklyProgressText = nil
            }

            let item = TodayTrackerItem(
                tracker: tracker,
                entry: todayEntry,
                weeklyProgressText: weeklyProgressText
            )

            if todayEntry == nil {
                pendingItems.append(item)
            } else {
                recordedItems.append(item)
            }
        }

        return TodayOverview(
            activeTrackerCount: activeTrackers.count,
            pendingItems: pendingItems,
            recordedItems: recordedItems
        )
    }

    static func buildTodayItems(
        trackers: [Tracker],
        entries: [TrackingEntry],
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday
    ) -> [TodayTrackerItem] {
        buildTodayOverview(
            trackers: trackers,
            entries: entries,
            dayKey: dayKey,
            weekStartsOn: weekStartsOn
        )
        .allItems
    }

    static func buildHistorySections(trackers: [Tracker], entries: [TrackingEntry]) -> [HistoryDaySection] {
        let trackerLookup = Dictionary(uniqueKeysWithValues: trackers.map { ($0.id, $0) })
        let grouped = Dictionary(grouping: entries, by: \.dayKey)

        return grouped
            .keys
            .sorted(by: >)
            .map { dayKey in
                let items = grouped[dayKey, default: []]
                    .sorted { $0.recordedAt > $1.recordedAt }
                    .compactMap { entry -> HistoryDayItem? in
                        guard let tracker = trackerLookup[entry.trackerId] else { return nil }

                        return HistoryDayItem(
                            trackerName: tracker.name,
                            colorToken: tracker.colorToken,
                            recordedAt: entry.recordedAt,
                            payloadSummary: entry.summary
                        )
                    }

                return HistoryDaySection(dayKey: dayKey, items: items)
            }
            .filter { !$0.items.isEmpty }
    }

    static func buildStatisticsSummary(
        trackers: [Tracker],
        entries: [TrackingEntry],
        today: DayKey,
        preferences: UserPreference
    ) -> StatisticsSummary {
        let activeTrackers = trackers.filter { !$0.isArchived }
        let allEntryCount = entries.count
        let windowStart = today.addingDays(-(preferences.preferredStatisticsWindow.dayCount - 1)) ?? today
        let completedKeys = Set(entries.filter(\.countsAsCompletion).map { key(trackerID: $0.trackerId, dayKey: $0.dayKey) })

        var dueCount = 0
        var completedCount = 0
        var currentDay = windowStart

        while currentDay <= today {
            for tracker in activeTrackers where trackerCreatedDayKey(for: tracker) <= currentDay {
                let weeklyCountBeforeDay = completionCountBeforeDay(
                    entries: entries,
                    trackerID: tracker.id,
                    dayKey: currentDay,
                    weekStartsOn: preferences.weekStartsOn
                )

                if tracker.schedule.isDue(on: currentDay, completedCountInWeek: weeklyCountBeforeDay) {
                    dueCount += 1
                    if completedKeys.contains(key(trackerID: tracker.id, dayKey: currentDay)) {
                        completedCount += 1
                    }
                }
            }

            guard let nextDay = currentDay.addingDays(1) else { break }
            currentDay = nextDay
        }

        let currentStreakDays = buildCurrentStreakDays(entries: entries, today: today)
        let trackerBreakdown = activeTrackers.map { tracker in
            TrackerCompletionSummary(
                trackerID: tracker.id,
                trackerName: tracker.name,
                colorToken: tracker.colorToken,
                totalEntryCount: entries.filter { $0.trackerId == tracker.id }.count
            )
        }
        .sorted {
            if $0.totalEntryCount == $1.totalEntryCount {
                return $0.trackerName < $1.trackerName
            }

            return $0.totalEntryCount > $1.totalEntryCount
        }

        let rate = dueCount == 0 ? 0 : Double(completedCount) / Double(dueCount)

        return StatisticsSummary(
            activeTrackerCount: activeTrackers.count,
            totalEntryCount: allEntryCount,
            currentWindowCompletedCount: completedCount,
            currentWindowDueCount: dueCount,
            currentWindowCompletionRate: rate,
            currentStreakDays: currentStreakDays,
            trackerBreakdown: trackerBreakdown
        )
    }

    private static func buildCurrentStreakDays(entries: [TrackingEntry], today: DayKey) -> Int {
        let completedDays = Set(entries.filter(\.countsAsCompletion).map(\.dayKey))
        var streak = 0
        var cursor = today

        while completedDays.contains(cursor) {
            streak += 1
            guard let previousDay = cursor.addingDays(-1) else { break }
            cursor = previousDay
        }

        return streak
    }

    private static func completionCountBeforeOrOnDay(
        entries: [TrackingEntry],
        trackerID: UUID,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday
    ) -> Int {
        completionCount(
            entries: entries,
            trackerID: trackerID,
            dayKey: dayKey,
            weekStartsOn: weekStartsOn,
            includeCurrentDay: true
        )
    }

    private static func completionCountBeforeDay(
        entries: [TrackingEntry],
        trackerID: UUID,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday
    ) -> Int {
        completionCount(
            entries: entries,
            trackerID: trackerID,
            dayKey: dayKey,
            weekStartsOn: weekStartsOn,
            includeCurrentDay: false
        )
    }

    private static func completionCount(
        entries: [TrackingEntry],
        trackerID: UUID,
        dayKey: DayKey,
        weekStartsOn: MarkerWeekday,
        includeCurrentDay: Bool
    ) -> Int {
        guard let bounds = weekBounds(containing: dayKey, weekStartsOn: weekStartsOn) else {
            return 0
        }

        return entries.filter { entry in
            guard entry.trackerId == trackerID else { return false }
            guard entry.dayKey >= bounds.start && entry.dayKey <= bounds.end else { return false }
            return includeCurrentDay ? entry.dayKey <= dayKey : entry.dayKey < dayKey
        }
        .count
    }

    private static func weekBounds(containing day: DayKey, weekStartsOn: MarkerWeekday) -> (start: DayKey, end: DayKey)? {
        guard let weekday = day.weekday else { return nil }
        let offset = (weekday.rawValue - weekStartsOn.rawValue + 7) % 7
        guard let start = day.addingDays(-offset),
              let end = start.addingDays(6) else {
            return nil
        }

        return (start, end)
    }

    private static func trackerCreatedDayKey(for tracker: Tracker) -> DayKey {
        DayKey(date: tracker.createdAt, timeZone: .current)
    }

    private static func key(trackerID: UUID, dayKey: DayKey) -> String {
        "\(trackerID.uuidString)|\(dayKey.rawValue)"
    }
}
